mito-ai/mito_ai/providers.py

```diff
@@ -1,70 +1,81 @@
+    CompletionItemError,
+    CompletionReply,
+    CompletionRequest,
+    CompletionStreamChunk,
+)
+from .utils.db import get_user_field, set_user_field
+from .utils.open_ai_utils import get_ai_completion_from_mito_server
+from .utils.schema import UJ_AI_MITO_API_NUM_USAGES
+from .utils.telemetry_utils import (
+    KEY_TYPE_PARAM,
+    MITO_AI_COMPLETION_ERROR,
+    MITO_AI_COMPLETION_SUCCESS,
+    MITO_SERVER_KEY,
+    MITO_SERVER_NUM_USAGES,
+    USER_KEY,
+    log,
 )
 
 __all__ = ["OpenAIProvider"]
-
-
-COMPLETION_SYSTEM_PROMPT = """
-You are an application built to provide helpful code completion suggestions.
-You should only produce code. Keep comments to minimum, use the
-programming language comment syntax. Produce clean executable code.
-The code is written for a data analysis and code development
-environment which can execute code to produce graphics, tables and
-interactive outputs.
-"""
-
-
-COMPLETION_DEFAULT_TEMPLATE = """
-The document is called `{{filename}}` and written in {{language}}.
-"""
+_num_usages = None
 
 
 class OpenAIProvider(LoggingConfigurable):
     """Provide AI feature through OpenAI services."""
-    # Internally it uses jinja2 template to render prompt messages.
 
     api_key = Unicode(
         config=True,
         help="OpenAI API key. Default value is read from the OPENAI_API_KEY environment variable.",
     )
 
+    max_completion_tokens = CInt(
+        None,
+        allow_none=True,
+        config=True,
+        help="An upper bound for the number of tokens that can be generated for a completion, including visible output tokens and reasoning tokens.",
+    )
+
     # FIXME add validate function to check if the model is valid
     model = Unicode(
         "gpt-4o-mini", config=True, help="OpenAI model to use for completions"
     )
 
+    temperature = CFloat(
+        0,
+        config=True,
+        help="What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
+    )
+
     def __init__(self, **kwargs) -> None:
-        super().__init__(**kwargs)
+        super().__init__(log=get_logger(), **kwargs)
         self._client: Optional[AsyncOpenAI] = None
-        # Load jinja2 templates
-        self._templates = Environment(
-            loader=DictLoader(
-                {
-                    "completion-system": COMPLETION_SYSTEM_PROMPT,
-                    "completion-human": COMPLETION_DEFAULT_TEMPLATE,
-                }
-            )
-        )
 
     @default("api_key")
     def _api_key_default(self):
         return os.environ.get("OPENAI_API_KEY", "")
 
+    @property
+    def can_stream(self) -> bool:
+        """Whether the provider supports streaming completions.
+
+        Streaming is only supported if an OpenAI API key is provided.
+        """
+        return bool(self.api_key)
+
     @property
     def client(self) -> AsyncOpenAI:
```

Comment:
I think that this property makes less sense now that we support using the `mito-ai-server` if no open ai key is set. Maybe we should update the return type to be `Union[AsyncOpenAI, None]` and have it `None` when no OpenAI Key is set. Then in the `request_completions` function we can just case on the `self.client is not None` instead of checking for the open ai key again. 

And maybe we can update the name of the property to `user_key_client` or something like that to make it more clear when this client is useful. 

What do you think?

Reply:
I would go for `_openAI_client` if it is ok for you - the reason behind that is that it is local to the provider.